Add message flow and PEP/PDP drill-down to basic scenario (#121)
jsalowey/wimse-arch: draft-ietf-wimse-arch.md

There can be variations on cross domain workflows. For example, in step 3 the workload was able to use its Workload Identity Credentials to directly access an infrastructure service. It also may be possible for an workload to request an access token for an external service using its Workload Identity Credentials directly with an external token service.

## Message Flow and Policy Enforcement in the Basic Scenario {#arch-basic-ops}

> The basic workload identity scenario in {{basic-workload-identity-scenario}} (see {{arch-basic}}) already sketches a trust domain, a gateway, a CA/credential service, and how an application client reaches multiple workloads. This section takes the same scenario and adds one level of detail for a single workload-to-workload hop: a caller and a callee, identity established with workload identity credentials, and a separate authorization step that may involve policy components.

Reviewer: ```suggestion
The basic workload identity scenario described in {{basic-workload-identity-scenario}} and illustrated in {{arch-basic}} already sketches a trust domain, a gateway, a CA/credential service, and how an application client reaches multiple workloads. This section takes the same scenario and adds one level of detail for a single workload-to-workload hop: a caller and a callee, identity established with workload identity credentials, and a separate authorization step that may involve policy components.
```

> In this view, a Policy Enforcement Point (PEP) is the logical function that allows a request to proceed, blocks it, or applies obligations, after (or as part of) authenticating the peer. A Policy Decision Point (PDP) is an optional component to which a PEP delegates policy evaluation. Deployments may implement PEP and PDP as separate services, co-locate them with a gateway or sidecar, or embed them in the receiving workload. This document does not define a particular policy language or protocol between PEP and PDP.

Reviewer: ```suggestion
In this view, the component enforcing authorization policy acts as a Policy Enforcement Point (PEP): it allows a request to proceed, blocks it, or applies obligations based on the authenticated peer identity and any relevant security context. A deployment may also use a Policy Decision Point (PDP), to which the enforcing component delegates policy evaluation. Deployments may implement PEP and PDP as separate services, co-locate them with a gateway or sidecar, or embed them in the receiving workload. This document does not define a particular policy language or protocol between PEP and PDP.
```

> The diagram below is at the same level of abstraction as the flows in the workload credentials document {{?I-D.ietf-wimse-workload-creds}}: it is not a normative description of a single product, but a common logical layout for how provisioning, connection setup, application-level authentication, authorization, and response relate when two workloads communicate.

Reviewer: ```suggestion
The diagram below is illustrative rather than normative. It shows one common logical layout for how provisioning, connection setup, application-level authentication, authorization, and response relate when two workloads communicate.
```

~~~aasvg
+------------+               +------------+
|            |      (2)      |            |
|            |<=============>|            |
|            |               |            |
| Workload A |      (3)      | Workload B |
|            |==============>|            |
|            |               |            |
|            |      (5)      |   +--------+
|            |<==============|   |  PEP   |
+------------+               +---+--------+
      ^                        ^     ^
      |            (1)         |     |
  (1) | +----------------------+     | (4)
      | |                            |
      v v                            v
+------------+               +------------+
|            |               |            |
|  CA/       |               |    PDP     |
|  Credential|               | (optional) |
|  Service   |               |            |
+------------+               +------------+
~~~
{: #arch-basic-ops-fig title="Logical message flow and policy points (drill-down of the basic scenario)."}

> The CA/credential service issues workload identity credentials; workloads obtain them early in the lifecycle (provisioning details vary by deployment). The PEP is shown on the callee (Workload B) path; alternatively it may also sit in a gateway or elsewhere, as in {{arch-basic}}.

Reviewer: ```suggestion
The CA/credential service issues workload identity credentials; workloads obtain them before using those credentials for authentication. Provisioning details vary by deployment. The PEP is shown as a logical function associated with the callee-side request path. It may be implemented inside Workload B, in a sidecar, in a gateway, or in another component on the request path, as in {{arch-basic}}.
```

The high-level message flow is as follows:

> 1. Workload A and Workload B each obtain a workload identity credential from the CA/credential service, typically well before a given application request. This may happen on the order of hours between refreshes.

Reviewer: ```suggestion
1. Workload A and Workload B each obtain a workload identity credential from the CA/credential service before using that credential for authentication. The timing, lifetime, and refresh behavior of these credentials are deployment-specific. Normally credentials operate on a slower lifecycle than application requests.
```

Author: I think your proposed text is a bit too general here. Specifically it omits to mention that credentials operate on a slower lifecycle then application requests.

Reviewer: I'd rather avoid specific time guidance here as such things are very much use case specific. How about the updated version?

Author: Things are use-case specific, but when you say "hours" it's clear to people that everything is automated. Otherwise, people might still have the old PKI model in mind.

In any case, your draft, your call.

2. A transport connection is set up for the call. It may use mutual TLS and workload identity certificates, or another transport security mechanism.

> 3. Workload A sends a request to Workload B. This may include application-level authentication using a Workload Identity Token and proof of possession, as defined in the credentials and protocol documents. Workload B authenticates Workload A.

Reviewer: ```suggestion
3. Workload A sends a request to Workload B. This may include application-level authentication using a Workload Identity Token and proof of possession, as defined in the credentials and protocol documents. Workload A is authenticated to Workload B, or to a component acting on Workload B's request path.
```

Author: The new text is confusing IMHO. Regardless of where the physical auth work gets done, workload B needs to know that it was successful, so effectively "Workload A is authenticated to Workload B" and the second part of that sentence is redundant.

Reviewer: Fair enough. Let's drop this suggestion.
